pkg/server/handler/auth.go

```diff
@@ -1056,3 +1057,115 @@ func (h *LinkProviderHandler) Handle(payload *router.Payload, response *router.R
 	response.Result = "OK"
 	return
 }
+
+//  Define the playload for sso plugin to connect user with provider
+type unlinkProviderPayload struct {
+	Provider            string                 `mapstructure:"provider"`
+	PrincipalIDPrefix   string
+	UserID              string                 `mapstructure:"user_id"`
+}
+
+func (payload *unlinkProviderPayload) Decode(data map[string]interface{}) skyerr.Error {
+	if err := mapstructure.Decode(data, payload); err != nil {
+		return skyerr.NewError(skyerr.BadRequest, "fails to decode the request payload")
+	}
+	payload.PrincipalIDPrefix = payload.Provider + ":"
+	return payload.Validate()
+}
+
+func (payload *unlinkProviderPayload) Validate() skyerr.Error {
+	if payload.Provider == "" {
+		return skyerr.NewInvalidArgument("empty provider", []string{"provider"})
+	}
+
+	if payload.UserID == "" {
+		return skyerr.NewInvalidArgument("empty user id", []string{"user_id"})
+	}
+
+	return nil
+}
+
+// UnlinkProviderHandler disconnect user with specific provider
+//
+// UnlinkProviderHandler receives parameters:
+//
+// * provider (string, required)
+// * user_id (string, required)
+//
+// curl -X POST -H "Content-Type: application/json" \
+//   -d @- http://localhost:3000/ <<EOF
+// {
+// 		"action": "auth:provider:unlink",
+// 		"provider": "facebook",
+// 		"user_id": "c0959b6b-15ea-4e21-8afb-9c8308ad79db"
+// }
+// EOF
+// Response
+// {
+//     "result": "OK"
+// }
+type UnlinkProviderHandler struct {
+	HookRegistry     *hook.Registry     `inject:"HookRegistry"`
+	AssetStore       asset.Store        `inject:"AssetStore"`
+	AuthRecordKeys   [][]string         `inject:"AuthRecordKeys"`
+	AccessKey        router.Processor   `preprocessor:"accesskey"`
+	DBConn           router.Processor   `preprocessor:"dbconn"`
+	InjectPublicDB   router.Processor   `preprocessor:"inject_public_db"`
+	PluginReady      router.Processor   `preprocessor:"plugin_ready"`
+	RequireMasterKey router.Processor   `preprocessor:"require_master_key"`
+	preprocessors    []router.Processor
+}
+
+func (h *UnlinkProviderHandler) Setup() {
+	h.preprocessors = []router.Processor{
+		h.AccessKey,
+		h.DBConn,
+		h.InjectPublicDB,
+		h.PluginReady,
+		h.RequireMasterKey,
+	}
+}
+
+func (h *UnlinkProviderHandler) GetPreprocessors() []router.Processor {
+	return h.preprocessors
+}
+
+func (h *UnlinkProviderHandler) Handle(payload *router.Payload, response *router.Response) {
+	log.Debugf("Unlink provider")
+	p := &unlinkProviderPayload{}
+	skyErr := p.Decode(payload.Data)
+	if skyErr != nil {
+		response.Err = skyErr
+		return
+	}
+
+	info := skydb.AuthInfo{}
+	userID := p.UserID
+
+	if err := payload.DBConn.GetAuth(userID, &info); err != nil {
+		response.Err = skyerr.NewError(skyerr.ResourceNotFound, "user not found")
+		return
+	}
+
+	providerInfo := info.ProviderInfo
+	unlinked := false
+	for k := range providerInfo {
+		if strings.HasPrefix(k, p.PrincipalIDPrefix) {
+			info.RemoveProviderInfoData(k)
+			unlinked = true
+		}
+	}
```

Comment:
`RemoveProviderInfoData` should return an error on fails, or no connected provider. Handler should not responsible for iteration and checking key existence.